go/arrow/array/list.go

```diff
@@ -618,19 +634,1061 @@ func (b *baseListBuilder) UnmarshalJSON(data []byte) error {
+
+func (a *LargeListView) ValueOffsets(i int) (start, end int64) {
+	debug.Assert(i >= 0 && i < a.array.data.length, "index out of range")
+	j := i + a.array.data.offset
+	size := a.sizes[j]
+	// If size is 0, skip accessing offsets.
+	if size == 0 {
+		return 0, 0
+	}
+	start = a.offsets[j]
+	end = start + size
+	return
+}
+
+func (a *LargeListView) Retain() {
+	a.array.Retain()
+	a.values.Retain()
+}
+
+func (a *LargeListView) Release() {
+	a.array.Release()
+	a.values.Release()
+}
+
+// Acessors for offsets and sizes to make ListView and LargeListView validation generic.
+type offsetsAndSizes interface {
+	offsetAt(slot int64) int64
+	sizeAt(slot int64) int64
+}
+
+var _ offsetsAndSizes = (*ListView)(nil)
+var _ offsetsAndSizes = (*LargeListView)(nil)
+
+func (a *ListView) offsetAt(slot int64) int64 { return int64(a.offsets[int64(a.data.offset)+slot]) }
+
+func (a *ListView) sizeAt(slot int64) int64 { return int64(a.sizes[int64(a.data.offset)+slot]) }
+
+func (a *LargeListView) offsetAt(slot int64) int64 { return a.offsets[int64(a.data.offset)+slot] }
+
+func (a *LargeListView) sizeAt(slot int64) int64 { return a.sizes[int64(a.data.offset)+slot] }
+
+func outOfBoundsListViewOffset(l offsetsAndSizes, slot int64, offsetLimit int64) error {
+	offset := l.offsetAt(slot)
+	return fmt.Errorf("%w: Offset invariant failure: offset for slot %d out of bounds. Expected %d to be at least 0 and less than %d", arrow.ErrInvalid, slot, offset, offsetLimit)
+}
+
+func outOfBoundsListViewSize(l offsetsAndSizes, slot int64, offsetLimit int64) error {
+	size := l.sizeAt(slot)
+	if size < 0 {
+		return fmt.Errorf("%w: Offset invariant failure: size for slot %d out of bounds: %d < 0", arrow.ErrInvalid, slot, size)
+	} else {
+		offset := l.offsetAt(slot)
+		return fmt.Errorf("%w: Offset invariant failure: size for slot %d out of bounds: %d + %d > %d", arrow.ErrInvalid, slot, offset, size, offsetLimit)
+	}
+}
+
+// Pre-condition: Basic validation has already been performed
+func (a *array) fullyValidateOffsetsAndSizes(l offsetsAndSizes, offsetLimit int64) error {
+	for slot := int64(0); slot < int64(a.Len()); slot += 1 {
+		size := l.sizeAt(slot)
+		if size > 0 {
+			offset := l.offsetAt(slot)
+			if offset < 0 || offset > offsetLimit {
+				return outOfBoundsListViewOffset(l, slot, offsetLimit)
+			}
+			if size > offsetLimit-int64(offset) {
+				return outOfBoundsListViewSize(l, slot, offsetLimit)
+			}
+		} else if size < 0 {
+			return outOfBoundsListViewSize(l, slot, offsetLimit)
+		}
+	}
+
+	return nil
+}
+
+func (a *array) validateOffsetsAndMaybeSizes(l offsetsAndSizes, offsetByteWidth int, isListView bool, offsetLimit int64, fullValidation bool) error {
+	nonEmpty := a.Len() > 0
+	if a.data.buffers[1] == nil {
+		// For length 0, an empty offsets buffer is accepted (ARROW-544).
+		if nonEmpty {
+			return fmt.Errorf("non-empty array but offsets are null")
+		} else {
+			return nil
+		}
+	}
+	if isListView {
+		if a.data.buffers[2] == nil {
```

Comment:
can we just combine these into a single conditional?

Reply:
Yes. This reads a bit weird because it comes from the C++ code that had `if constexpr` on the outermost `if`.